### PR TITLE
DEPR: Silent dropping of nuisance columns in transform_dict_like

### PR DESCRIPTION
**Pull request by rhshadrach**
Part of #43740

- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -337,6 +337,7 @@ Other Deprecations
 - Deprecated the ``squeeze`` argument to :meth:`read_csv`, :meth:`read_table`, and :meth:`read_excel`. Users should squeeze the DataFrame afterwards with ``.squeeze("columns")`` instead. (:issue:`43242`)
 - Deprecated the ``index`` argument to :class:`SparseArray` construction (:issue:`23089`)
 - Deprecated :meth:`.Rolling.validate`, :meth:`.Expanding.validate`, and :meth:`.ExponentialMovingWindow.validate` (:issue:`43665`)
+- Deprecated silent dropping of columns that raised a ``TypeError`` in :class:`Series.transform` and :class:`DataFrame.transform` when used with a dictionary (:issue:`43740`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.4.0.rst:340]: is GroupBy.transform also affected?  (same question for the agg PR)

Comment (rhshadrach) [reply doc/source/whatsnew/v1.4.0.rst:340]: No - groupby.transform does not accept dictionaries (or lists)

**Review by jreback — APPROVED** [review]
lgtm

**Comment by jbrockmendel** [comment]
@rhshadrach want to make a PR enforcing this deprecation?

**Comment by rhshadrach** [comment]
Yep, will do.

**Pull request merged**